### PR TITLE
Add SparsityCsr sorting kernels

### PR DESCRIPTION
**Pull request by upsj**
and fallback routines for missing int64 sparselib implementations

### CODE REVIEW

File: test/matrix/sparsity_csr_kernels.cpp [hunk 1]
```diff
@@ -145,4 +147,100 @@ TEST_F(SparsityCsr, ConvertToDenseIsEquivalentToRef)
...
+    gko::matrix_data<value_type, index_type> data;
+    gko::matrix_data<value_type, gko::int64> data64;
+    mtx->sort_by_column_index();
+    mtx->write(data);
+    data64.size = data.size;
+    for (auto entry : data.nonzeros) {
+        data64.nonzeros.emplace_back(entry.row, entry.column, entry.value);
+    }
+    mtx64->read(data64);
+    dmtx64->read(data64);
```
Comment (yhmtsai) [inline test/matrix/sparsity_csr_kernels.cpp:219]: it's not unsorted matrix

**Review by yhmtsai — CHANGES_REQUESTED** [review]
only for the test. others LGTM.

File: test/matrix/sparsity_csr_kernels.cpp (hunk 1 above)
Comment (upsj) [reply test/matrix/sparsity_csr_kernels.cpp:219]:
```suggestion
    mtx->write(data);
    data64.size = data.size;
    for (auto entry : data.nonzeros) {
        data64.nonzeros.emplace_back(entry.row, entry.column, entry.value);
    }
    mtx64->read(data64);
    dmtx64->read(data64);
```

File: core/test/utils/unsort_matrix.hpp [hunk 2]
```diff
@@ -55,24 +56,10 @@ namespace test {
...
-    using index_type = gko::detail::pointee<decltype(mtx->get_col_idxs())>;
-    auto nnz = mtx->get_num_stored_elements();
-    if (nnz <= 0) {
-        return;
-    }
-
+    using value_type = typename gko::detail::pointee<MtxPtr>::value_type;
+    using index_type = typename gko::detail::pointee<MtxPtr>::index_type;
     const auto exec = mtx->get_executor();
     const auto master = exec->get_master();
```
Comment (MarcelKoch) [inline core/test/utils/unsort_matrix.hpp:62]: unused
```suggestion
```

File: dpcpp/matrix/sparsity_csr_kernels.dp.cpp [hunk 3]
```diff
@@ -303,7 +303,23 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
+    auto queue = exec->get_queue();
+    // build sorted postorder node list for each row
+    queue->submit([&](sycl::handler& cgh) {
+        cgh.parallel_for(sycl::range<1>{num_rows}, [=](sycl::id<1> idx_id) {
+            const auto row = idx_id[0];
+            const auto row_begin = row_ptrs[row];
+            const auto row_end = row_ptrs[row + 1];
+            // heap-sort the elements
+            std::make_heap(cols + row_begin, cols + row_end);
+            std::sort_heap(cols + row_begin, cols + row_end);
```
Comment (MarcelKoch) [inline dpcpp/matrix/sparsity_csr_kernels.dp.cpp:319]: I guess the reason you use heap sort here is that std::sort is not available in kernels?

**Review by MarcelKoch — APPROVED** [review]
LTGM. Only one very small nit.

File: dpcpp/matrix/sparsity_csr_kernels.dp.cpp (hunk 3 above)
Comment (upsj) [reply dpcpp/matrix/sparsity_csr_kernels.dp.cpp:319]: heapsort doesn't need recursion (at least the libstdc++ version is unrolled), while std::sort may use recursive calls.

File: dpcpp/matrix/sparsity_csr_kernels.dp.cpp [hunk 4]
```diff
@@ -312,8 +328,30 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
 template <typename ValueType, typename IndexType>
 void is_sorted_by_column_index(
     std::shared_ptr<const DpcppExecutor> exec,
-    const matrix::SparsityCsr<ValueType, IndexType>* to_check,
-    bool* is_sorted) GKO_NOT_IMPLEMENTED;
+    const matrix::SparsityCsr<ValueType, IndexType>* to_check, bool* is_sorted)
+{
+    array<bool> is_sorted_device_array{exec, {true}};
```
[the remark this answers is not in the document]
Comment (upsj) [reply dpcpp/matrix/sparsity_csr_kernels.dp.cpp:333]: done

**Review by yhmtsai — APPROVED** [review]
LGTM

**Comment by upsj** [comment]
rebase!

**Comment by upsj** [comment]
Dpcpp not working correctly so far

**Comment by upsj** [comment]
rebase!

**Pull request merged**